src/irmin/tree_intf.ml

```diff
@@ -0,0 +1,288 @@
+
+  type t
+  (** The type of trees. *)
+
+  val t : t Type.t
+  (** The value-type of trees. *)
+
+  (** {1 Constructors} *)
+
+  val empty : t
+  (** [empty] is the empty tree. The empty tree does not have associated backend
+      configuration values, as they can perform in-memory operation,
+      independently of any given backend. *)
+
+  val of_contents : ?metadata:metadata -> contents -> t
+  (** [of_contents c] is the subtree built from the contents [c]. *)
+
+  val of_node : node -> t
+  (** [of_node n] is the subtree built from the node [n]. *)
+
+  val v : [< `Node of node | `Contents of contents * metadata ] -> t
+  (** General-purpose constructor for trees. *)
+
+  val destruct : t -> [> `Node of node | `Contents of contents * metadata ]
+  (** General-purpose destructor for trees. *)
+
+  val kind : t -> key -> [ `Contents | `Node ] option Lwt.t
+  (** [kind t k] is the type of [s] in [t]. It could either be a tree node or
+      some file contents. It is [None] if [k] is not present in [t]. *)
+
+  val list : t -> key -> (step * [ `Contents | `Node ]) list Lwt.t
+  (** [list t key] is the list of files and sub-nodes stored under [k] in [t]. *)
+
+  (** {1 Diffs} *)
+
+  val diff : t -> t -> (key * (contents * metadata) Diff.t) list Lwt.t
+  (** [diff x y] is the difference of contents between [x] and [y]. *)
+
+  (** {1 Manipulating Contents} *)
+
+  val mem : t -> key -> bool Lwt.t
+  (** [mem t k] is true iff [k] is associated to some contents in [t]. *)
+
+  val find_all : t -> key -> (contents * metadata) option Lwt.t
+  (** [find_all t k] is [Some (b, m)] if [k] is associated to the contents [b]
+      and metadata [m] in [t] and [None] if [k] is not present in [t]. *)
+
+  val find : t -> key -> contents option Lwt.t
+  (** [find] is similar to {!find_all} but it discards metadata. *)
+
+  val get_all : t -> key -> (contents * metadata) Lwt.t
+  (** Same as {!find_all} but raise [Invalid_arg] if [k] is not present in [t]. *)
+
+  val get : t -> key -> contents Lwt.t
+  (** Same as {!get_all} but ignore the metadata. *)
+
+  val add : t -> key -> ?metadata:metadata -> contents -> t Lwt.t
+  (** [add t k c] is the tree where the key [k] is bound to the contents [c] but
+      is similar to [t] for other bindings. *)
+
+  val remove : t -> key -> t Lwt.t
+  (** [remove t k] is the tree where [k] bindings has been removed but is
+      similar to [t] for other bindings. *)
+
+  (** {1 Manipulating Subtrees} *)
+
+  (** Operations on {{!node} tree nodes}. *)
+  module Node : sig
+    type t = node
+    (** The type of tree nodes. *)
+
+    val t : t Type.t
+    (** The value type of tree nodes. *)
+
+    val hash : t -> hash
+    (** [hash t] is the hash of [t]. *)
+
+    val list : t -> (step * [ `Contents | `Node ]) list Lwt.t
+    (** [list t] is the list of keys in [t], and their corresponding kinds. *)
+
+    val bindings :
+      t -> (step * [ `Node of t | `Contents of contents * metadata ]) list Lwt.t
```

Comment:
I'm wondering if we should keep the content hash here, so that'll would avoid throwing `Corrupted_store` exception for shallow objects. Maybe returning `[ `Node of t | `Contents of hash * metadata ]` instead? Or even [ `Node of t | `Contents of hash * contents option * metadata ]` What do you think?

Reply:
Your point made me wonder about the analogous case for lazy nodes that don't exist in the store either. It seems we just return the empty list in such cases (https://github.com/CraigFe/irmin/blob/tree-bindings/src/irmin/tree.ml#L674-L675), which doesn't seem like the right behaviour to me.

Returning a hash works for the `contents` case, but I think either a `result` or exception is needed for the node case. What do you think?

Reply:
yes returning an empty list seems a bit odd here too. Returning a `result` makes sense.